src/PreconditionerTypes.f90

```diff
@@ -343,6 +550,699 @@ SUBROUTINE setup_ILU_PreCondtype(thisPC)
+      SELECT TYPE(LU => thisPC%LU(k)); TYPE IS(DenseSquareMatrixType)
+        CALL doolittle_LU_RSOR(A%blocks(k),LU)
+      ENDSELECT
+    ENDDO
+    ! blocks do not need to be set to zero because we can set the block mask
+  CLASS DEFAULT
+    !setup the Upper and Lower portion of the diagonal
+    DO k=1,thisPC%nLocalBlocks
+      DO i=1,thisPC%blockSize
+        DO j=1,thisPC%blockSize
+          CALL A%get((k+thisPC%blockOffset-1)*thisPC%blockSize+i, &
+              (k+thisPC%blockOffset-1)*thisPC%blockSize+j,tempreal)
+          CALL thisPC%LU(k)%set(i,j,tempreal)
+          IF(tempreal .NE. 0.0_SRK)THEN
+            CALL thispC%LpU%set((k+thisPC%blockOffset-1)*thisPC%blockSize+i, &
+                (k+thisPC%blockOffset-1)*thisPC%blockSize+j,0.0_SRK)
+          ENDIF
+        ENDDO
+      ENDDO
+    ENDDO
+    !do LU factorization on the diagonal blocks
+    DO k=1,thisPC%nLocalBlocks
+      SELECT TYPE(mat => thisPC%LU(k))
+      CLASS IS(DenseSquareMatrixType)
+        CALL doolittle_LU_RSOR(mat)
+      ENDSELECT
+    ENDDO
+  ENDSELECT
+ENDSUBROUTINE setup_DistributedRSOR_PreCondtype
+!
+!-------------------------------------------------------------------------------
+!> @brief Applies up the RSOR Preconditioner Type with a parameter list
+!> @param thisPC The preconditioner to act on
+!> @param v The vector to apply the preconditioner to to act on
+!>
+SUBROUTINE apply_DistributedRSOR_PreCondType(thisPC,v)
+  CLASS(DistributedRSOR_PrecondType),INTENT(INOUT) :: thisPC
+  CLASS(Vectortype),ALLOCATABLE,INTENT(INOUT) :: v
+  CHARACTER(LEN=*),PARAMETER :: myName='apply_DistributedRSOR_PreCondType'
+  TYPE(NativeDistributedVectorType)::w
+  TYPE(ParamType)::PListVec_RSOR
+  INTEGER(SIK)::k,lowIdx,highIdx
+
+  REQUIRE(thisPC%isInit)
+  REQUIRE(ALLOCATED(v))
+  REQUIRE(v%isInit)
+
+  SELECT TYPE(v)
+  CLASS IS(NativeDistributedVectorType)
+    CALL PListVec_RSOR%add('VectorType->n',thisPC%A%n)
+    CALL PListVec_RSOR%add('VectorType->chunkSize',thisPC%blockSize)
+    CALL PListVec_RSOR%add('VectorType->MPI_Comm_ID',thisPC%comm)
+    CALL PListVec_RSOR%add('VectorType->nlocal',SIZE(v%b))
+    CALL w%init(PListVec_RSOR)
+    w%b=v%b
+
+    DO k=1,thisPC%nLocalBlocks
+      SELECT TYPE(mat => thisPC%LU(k))
+      CLASS IS(DenseSquareMatrixType)
+        lowIdx = (k - 1)*thisPC%blockSize + LBOUND(w%b,1)
+        highIdx = lowIdx + thisPC%blockSize-1
+        ! This is not a matvec call, but a call to the triangular solver
+        ! dtrsv_all
+        CALL BLAS_matvec('L','N','U',mat%A,w%b(lowIdx:highIdx))
+        CALL BLAS_matvec('U','N','N',mat%A,w%b(lowIdx:highIdx))
+      ENDSELECT
+    ENDDO
+
+    SELECT TYPE(LpU => thisPC%LpU); TYPE IS(DistributedBlockBandedMatrixType)
+      CALL LpU%setBlockMask(.TRUE.)
+    ENDSELECT
+
+    WRITE(*,*) "w",w%b
+    WRITE(*,*) "v",v%b
+    CALL BLAS_matvec(THISMATRIX=thisPC%LpU,X=w,Y=v,BETA=1.0_SRK,&
+                      ALPHA=-thisPC%omega)
+    WRITE(*,*) "w",w%b
+    WRITE(*,*) "v",v%b
```

Comment:
These checking write statements should probably be removed.

Reply:
Cool, these will be removed. My mistake.
The RSORPrecon and RSORPreconParallel Unit tests should probably remain separate, because one is designed to be called with MPI and the other is not. So I think that will stay the same as it always has.